009-ui-and-admin-server-security.md

```diff
@@ -0,0 +1,84 @@
+
+The outlined proposal relies on Kafka to perform authentication or authorization and the Admin Server itself does not.
+In a Kafka environment, this may be acceptable but it is plausible that a UI and Admin Server may wish to communicate with non-Kafka environments in the future like the Kube API.
+In these environments authentication and authorization can again be delegated to the backend system.
+A discussion and agreement is needed whether we wish to implement authentication and authorization on the Admin Server itself.
+
+Proceeding with the delegated model, I have 3 sets of diagrams showing a SCRAM-SHA-512 solution, an OAuth2 solution and a Mutual TLS solution with the last one only being relevant to the Admin Server.
+
+## SCRAM-SHA-512
+![Kafka Client using SCRAM-SHA512](./images/scram-sha512-client.png)
+
+Strimzi supports SCRAM-SHA512 using native Kafka facilities, so the Kafka client is setup with credentials for the user and these credentials are passed to the broker where the broker authenticates against user credentials stored in Zookeeper.
+When the request is authenticated, the Principal is passed to a Kafka Simple ACL Authorizer which determines whether the user is authorized to access the resource.
+Access is defined through ACLs and are stored on Zookeeper.
+
+![Strimzi UI using SCRAM-SHA512](./images/scram-sha512-ui.png)
+
+This Kafka client model translates to a UI and Admin Server in a similar way.
+When the user loads up the UI in a browser, they are presented with a dialogue box requiring them to enter a userId and password.
+The userId and password they enter are pre-defined to Kafka by the creation of a Kafka user through the `User Operator` or the Kafka `bin/kafka-configs.sh ` command line tool.
+Once the userId and password have been entered, the UI validates the credentials on Zookeeper using `passport.js`.
+The result that is returned will be marked as valid or invalid.
+If the credentials are valid, they are stored in the session data associated with that browser session.
+When the UI requires data from the Admin Server, the credentials are recovered from the session data and placed as an auth header on the request.
+The Admin Server receives the request and configures the Kafka client in preparation for calling the Kafka broker.
+The SCRAM credentials are retrieved from the auth header and are added to the Kafka client configuration properties before executing the call to the Kafka broker.
+From thereon, the Kafka authentication and authorization mechanism works in an identical way to the description above for the Kafka Clients.
+
+This mechanism should be straightforward to setup using a `passport.js` strategy.
+The Strimzi UI is served from an Express server and that Express server will be inside the Strimzi cluster allowing the Express Server to talk to the Zookeeper on the secure port 2181.
```

Comment:
Ok, if this is done from the server, it can have Zookeeper access if deployed as part of the Kafka cluster. But it is still not clear to me how will you validate the scrambled credentials stored in Zookeeper. Also as I mentioned above, the timing is a bit unfortunate with Zookeeper being there today but gone probably in few months. So in the future this would need to talk with Kafka.

Reply:
I don't think that it makes sense to set up a ZooKeeper based mechanism right now when things are going to change in the future, that at this point should not be so far.
Maybe the UI can go through the admin server for authenticating the user, setting up a Kafka client for doing a "generic" request (i.e. get Kafka cluster-info) leveraging this request to check if authentication failed or not.

Reply:
These comments suggest we're not yet all on the same page. In case it's not already obvious:

* KIP-500 will remove ZK from the picture entirely. If it's not there you can't go sneakily looking in there for SCRAM-SHA hashes.
* Under the KIP-500 umbrella, KIP-554 defines the API which will be available for setting the SCRAM-SHA credentials from an `Admin` client instance. In a ZK-less cluster the broker will perform the hashing and write the hashes to the metadata log, defined in [KIP-631](https://cwiki.apache.org/confluence/display/KAFKA/KIP-631%3A+The+Quorum-based+Kafka+Controller#KIP631:TheQuorumbasedKafkaController-RecordFormats). and [KIP-595](https://cwiki.apache.org/confluence/display/KAFKA/KIP-595%3A+A+Raft+Protocol+for+the+Metadata+Quorum). Last I look the metadata log. 595 contains this:

    > At a high level, this proposal treats the metadata log as a topic with a single partition. The benefit of this is that many of the existing APIs can be reused. For example, it allows users to use a consumer to read the contents of the metadata log for _debugging purposes_. 

    (my emphasis). So while it might be possible, post KIP-500 to obtain SCRAM-SHA credientials it would not really be considered best practice from a Kafka PoV. 
* Therefore, we should try really hard to avoid needing to access to the SCRAM-SHA info.
* @ppatierno as mentioned above, using the provided credentials to make an authenticated connection to a broker works in theory, but since you can't tell that the broker actually used the credentials it relies on the whole thing being configured correctly to actually work. In the past someone suggested a `whoAmI()` method on the `Admin` client which, I guess could provide information about how the user was authenticated as well as who they were.

Reply:
So the summary above seems to be that we need to stay away from talking to Zookeeper. As we would want any UI we create to validate credentials that a user enters, we could potentially issue a call such as `describe cluster` to Kafka from an admin client and test whether it failed authentication. This assumes that the admin client will return an `AuthenticationException` if the credentials are invalid. I am not sure I quite understand why you would not know if the broker used the credentials in this case so I need some help in that. This mechanism seems very artificial but could potentially work. As the Admin Server would be stateless and therefore does not require a login process in order to establish a session, how limiting would it be to say that the UI only supports OAuth2 and SCRAM is restricted to the Admin Server?

Reply:
> So the summary above seems to be that we need to stay away from talking to Zookeeper. As we would want any UI we create to validate credentials that a user enters, we could potentially issue a call such as describe cluster to Kafka from an admin client and test whether it failed authentication. This assumes that the admin client will return an AuthenticationException if the credentials are invalid. I am not sure I quite understand why you would not know if the broker used the credentials in this case so I need some help in that. This mechanism seems very artificial but could potentially work.

Honestly, I'm not sure I see any other way. You could just not validate it at all and show the errors later inside the UI if the credentials don't work. There is no other way to check it. 

I assume that once ZooKeeper is removed, you would still be able to read the credentials somewhere from Kafka as you would today from ZooKeeper. But I think the logic for verifying the password will still be non-trivial. Plus it would not solve other mechanisms.

> how limiting would it be to say that the UI only supports OAuth2 and SCRAM is restricted to the Admin Server?

I think it is fine to decide that some mechanisms are not supported for this. The current TLS authentication has anyway its limitations in not supporting any revocation lists.

Do we want to also support some mode without any authentication? It is obviously not "production ready". But lot of people seem to start that way and it seems like it would be useful to have that mode as well.

Reply:
Sorry @scholzj I had always assumed we would have a `security off` mode so will add something about that to the doc. @tombentley , @ppatierno , @mstruk , @samuel-hawker can I get an opinion from you on whether we should drop SCRAM for the UI and support no-auth and OAuth2 only? We can continue to support SCRAM for the Admin Server.

Reply:
Supporting OAuth2 only means that the user has to deploy an OAuth server so one more component for the UI. 
Maybe he doesn't want to do that because he is happy with using SCRAM.
About doing a "test" call (i.e. `describe cluster`) for SCRAM authentication, it sounds to me that we agree that it could work but @tombentley not. Can you elaborate a little bit more the why, Tom?

Reply:
On second thoughts, I think it probably does work.

1. You configure an `Admin` client with some (list of) addresses and some SCRAM credentials 
2. You `describeCluster()` and you get some result.
3. Have you proved that the credentials you provided were valid?

Only if you know that the client actually connected using SCRAM. I guess if you've set the `security.protocol=SASL_*` and `sasl.mechanism` then that's probably a valid assumption, but I think it's worth having a test for.